cpp/log/cert.cc

```diff
@@ -546,12 +588,311 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
       LOG(WARNING) << "Corrupt extension data";
       LOG_OPENSSL_ERRORS(WARNING);
     }
+
     return FALSE;
   }
 
   return TRUE;
 }
 
+bool IsRedactedHost(const string& hostname) {
+  // Split the hostname on '.' characters
+  const vector<string> tokens(util::split(hostname, '.'));
+
+  for (const string& str : tokens) {
+    if (str == "?") {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool IsValidRedactedHost(const string& hostname) {
+  // Split the hostname on '.' characters
+  const vector<string> tokens(util::split(hostname, '.'));
+
+  // Enforces the following rules: '?' must be to left of non redactions
+  // If first label is '*' then treat it as if it was a redaction
+  bool can_redact = true;
+  for (int pos = 0; pos < tokens.size(); ++pos) {
+    if (tokens[pos] == "?") {
+      if (!can_redact) {
+        return false;
+      }
+    } else {
+      // Allow a leading '*' for redaction but once we've seen anything else
+      // forbid further redactions
+      if (tokens[pos] != "*") {
+        can_redact = false;
+      } else if (pos > 0) {
+        // '*' is only valid at the left
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+Cert::Status Cert::IsValidWildcardRedaction() const {
+  if (!IsLoaded()) {
+    LOG(ERROR) << "Cert not loaded";
+    return Cert::ERROR;
+  }
+
+  string common_name;
+  vector<string> dns_alt_names;
+  int redacted_name_count = 0;
+
+  // First. Check all the Subject Alt Name extension records. Any that are of
+  // type DNS must pass validation if they are attempting to redact labels
+  STACK_OF(GENERAL_NAME)* subject_alt_names = (STACK_OF(GENERAL_NAME)*)
```

Comment:
The compiler *should* be able to see through the expansions, but if this doesn't work, don't worry too much, I'd say...

```C++
unique_ptr<STACK_OF(GENERAL_NAME), void(STACK_OF(GENERAL_NAME)*)> subject_alt_names((STACK_OF(GENERAL_NAME)*)X509_get_ext_d2i(x509_, NID_subject_alt_name, NULL, NULL), sk_GENERAL_NAME_free);
```

I'll leave reformatting to you. :wink: 

Also, you've used a C-style cast, should use a C++-style one (probably `static_cast`, I'd guess here).